### ISSUE TITLE
How to deploy the server manually?

### ISSUE THREAD

**Issue by ramsestom** [playtomic/apiserver#12]
What are the steps I shoudl follow manually deploy the server?
The instructions in here: http://playtomic.org/server.html are only for an Heroku installation and I want to put it on openshift...
So once I have downloaded and copied the git source on my openshift application, what am I supposed to do? Is there any config file I must change or something?

**Comment by benlowry** [comment]
Hi ramsestom, to deploy it manually you need to set up the dependencies:

- nodes
- mongodb 

You'll need to create a database and user account in mongodb, and create credentials for your game just like the instructions here: http://playtomic.org/games.html except in your own mongodb database.

You will need an ENV var that contains your mongodb connection string:
`export playtomic=mongodb://..........`

And then it's just a nodejs project you can start via: 
`node web.js`

You may also set up mailchimp's ENV vars for newsletters, you can see the expected variables here:
https://github.com/playtomic/apiserver/blob/master/api/config.js

**Comment by ramsestom** [comment]
Thanks for your reply.
But I think I aslo need to change the node.js port in the web.js script (process.env.PORT does not exists on openshift so I guess it must be replaced by process.env.OPENSHIFT_NODEJS_PORT). Correct me if I am wrong.
I don't know if I need to define the server_ip_address (process.env.OPENSHIFT_NODEJS_IP) somwhere too...
As for the mongodb connection, I think I will modify it in the config.js file rather than putting it in an env variable...
To start the server, I defined : "main": "web.js" in package.json so it should be OK
 I don't know if I need to make any other changes...

Also, I guess I can delete the .c9 and .idea directories so as the Procfile and mongod file?

**Comment by benlowry** [comment]
You can change the port by providing a PORT var or changing the default here:
https://github.com/playtomic/apiserver/blob/master/web.js#L53

Those files should be deletable - c9 is from the web IDE, .idea is from IntelliJ I think, the Procfile is for starting the server on Heroku, and the mongod file is for starting mongodb on C9 (that was me).

**Comment by ramsestom** [comment]
Well it do not work
I think it as to do to the fact that, additionaly to the specific port, the application has to be bind to the IP address defined in process.env.OPENSHIFT_NODEJS_IP in order to recieve traffic from the routing layer (but I do not see anywhere to define it in the playtomic code. Only the Port can be changed, mot the IP...).
I never used node.js before so this is pretty obscure to me for now but I think I will start with the server.js provided by openshift sample node.js application and try to integrate part of the web.js code of playtomic in it...

**Comment by benlowry** [comment]
You might have more luck now, there was a pending pull request that added support for a module (express) that had obsoleted some parts of their API.

I have also prepared instructions for deploying to DigitalOcean:

create ubuntu droplet and ssh into server

# dependencies
$ sudo apt-get update && sudo apt-get install -y nodejs npm mongodb git
$ ln -s /usr/bin/nodejs /usr/bin/node

# copy api server and install its dependencies
$ git clone https://github.com/playtomic/apiserver.git
$ cd apiserver
$ npm install

# setup mongodb
$ mongo
use playtomic
db.addUser(‘playtomic’, ‘playtomic’)
exit

# setup config
echo PORT=3000 > start
echo playtomic=mongodb://playtomic:playtomic@127.0.0.1:27017/playtomic >> start
echo node web.js >> start

# start server
bash start

**Comment by ramsestom** [comment]
Yes seems to work now (the application build successfully on Openshift. Not tried to interact with it yet though). Thanks

And just for information. You have twice the same process.env.MONGOLAB_URI test in your config.js file now. So you can delete one ;)

**Comment by benlowry** [comment]
The keys date back to when playtomic was a commercial service, the public key was used to isolate each game to separate subdomains on the api server while the private key was the actual credential.  

They definitely shouldn't be ssh keys, you can use any value you want for them.

**Comment by Lucidarme** [comment]
Hello, I try to set up Playtomic on my own server too. I followed your tutorial and when I run `bash start`, I have the following errors:

> { [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
(node) sys is deprecated. Use util instead.
Listening on 3000, env: undefined, local: false

And then the server is listening, but when I try to connect a player, I always have the error 1, which is: 

> Unable to login: General error, this typically means the player is unable to connect to the server

I have searched for the error about module bson, and some people said to update mongoose. I tried to do that but it doesn't change anything. Have you an idea to resolve this error? Or perhaps I am wrong and this error is not important, and the problem comes to my authentication ?
Thanks for help
